src/Moryx.ControlSystem/Cells/Cell.cs

```diff
@@ -56,16 +145,55 @@ public void PublishReadyToWork(ReadyToWork readyToWork)
         public event EventHandler<ReadyToWork> ReadyToWork;
 
         /// <summary>
-        /// Publish a <see cref="NotReadyToWork"/> from the resource
+        /// Publish a <see cref="NotReadyToWork"/> from the resource.
+        /// If the session was started within <see cref="PublishReadyToWorkAsync(Moryx.ControlSystem.Cells.ReadyToWork)"/> that async call is canceled
         /// </summary>
         public void PublishNotReadyToWork(NotReadyToWork notReadyToWork)
         {
+            if (_sessionCompletionSources.TryGetValue(notReadyToWork.Id, out var completionSource))
+            {
+                _sessionCompletionSources.Remove(notReadyToWork.Id);
+                completionSource.TrySetCanceled();
+                // cancellation of the completionSource will send a NotReadyToWork 
+                return;
+            }
             NotReadyToWork?.Invoke(this, notReadyToWork);
         }
 
         /// <inheritdoc />
         public event EventHandler<NotReadyToWork> NotReadyToWork;
 
+        /// <summary>
+        /// Publish <see cref="ActivityCompleted"/> from the resource
+        /// </summary>
+        /// <param name="activityResult"></param>
+        public Task<Session> PublishActivityCompletedAsync(ActivityCompleted activityResult)
+        {
+            return PublishActivityCompletedAsync(activityResult, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Publish <see cref="ActivityCompleted"/> from the resource
+        /// </summary>
+        /// <param name="activityResult"></param>
+        /// <param name="cancellationToken"></param>
+        public Task<Session> PublishActivityCompletedAsync(ActivityCompleted activityResult, CancellationToken cancellationToken)
+        {
+            Logger.Log(LogLevel.Trace, $"PublishActivityCompletedAsync Session {activityResult.Id}, Classification {activityResult.AcceptedClassification}, {activityResult.Reference}");
+            using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(LifeCycleTokenSource.Token, cancellationToken);
+            var linkedToken = linkedTokenSource.Token;
+            // throw exception if cancellation via token was requested
+            linkedToken.ThrowIfCancellationRequested();
+            // double check if any of the combined Tokens are already canceled.
+            cancellationToken.ThrowIfCancellationRequested();
+            LifeCycleTokenSource.Token.ThrowIfCancellationRequested();
+
+            var completionSource = new TaskCompletionSource<Session>();
+            _sessionCompletionSources.Add(activityResult.Id, completionSource);
```

Comment:
The ConcurrentDictionary has a `TryAdd` method, still we need to decide here what happens when application code falsely tries to publish an activity completed twice. If there should be an exception this should be provided in the comment; otherwise we could use the `AddOrUpdate` method and send it to the process engine as would be done with the not-async method